L80-csharp-load-balancing.md

```diff
@@ -0,0 +1,259 @@
+* Load balancer name in service config will be used to create a load balancer.
+  * A collection of load balancer factories are fetched from the service provider. One is selected using the configured name (if none specified then `pick_first` is used) and it will create load balancer.
+  * Load balancer is initialized with an `IChannelController` instance by the channel. This is used to call back to the channel to create sub-channels and update the picker.
+
+### Connecting
+
+* No network activity until `GrpcChannel.ConnectAsync` is called, or the first RPC is started.
+* On connect, the `ConnectionManager` will subscribe to address resolver and call the resolver's `RefreshAsync` to prompt the resolver to run immediately.
+* Resolver will push addresses to `ConnectionManager` which will then call balancer.
+* Balancer will use `IChannelController` to create, remove or update sub-channels as needed based on the addresses.
+  * A new sub-channel will start connecting when `SubChannel.ConnectAsync` is called. A sub-channel will be shutdown when disposed.
+* Balancer will be notified on changes in sub-channel state. It will create a `SubChannelPicker` based on the current state.
+
+### Picking
+
+* `ConnectionManager.PickAsync` is called to get a ready sub-channel along with the sub-channel's connected address.
+* If a picker hasn't been set yet, or the picker is empty, or the picker has an error then `PickAsync` will wait for a picker with a ready sub-channel if `WaitForReady` has been set. If `WaitForReady` is not set then the error will be thrown back to the caller.
+
+## Rationale
+
+* Design shares the most in common with grpc-go. SubChannels have a collection of addresses and are responsible for staying connected and retrying when in TRANSIENT_FAILURE.
+* Goal is to support basic gRPC load balancing using DNS in Kubernetes.
+* Secondary goal is to have a design in place that can be used to implement xDS in the future with minimal changes.
+* All load balancing APIs will be marked as experimental.
+* API design of resolver and load balancer purposefully choses not to use static configuration. This is considered an anti-pattern in .NET. Instead custom resolver and load balancer factories are registered with a DI container which is then hooked up to the gRPC channel using `GrpcChannelOptions.ServiceProvider`.
+
+## Implementation
+
+Implementation is in the grpc-dotnet repo: [PR #1286](https://github.com/grpc/grpc-dotnet/pull/1286).
+
+```csharp
+namespace Grpc.Net.Client.Balancer {
+    public sealed class AddressResolverOptions {
+        public AddressResolverOptions();
+    }
+    public class SubChannelOptions {
+        public SubChannelOptions(IReadOnlyList<DnsEndPoint> addresses);
+        public IReadOnlyList<DnsEndPoint> Addresses { get; }
+    }
+    public interface IConnectionController {
+        ConnectivityState State { get; }
```

Comment:
My idea here was to store the connectivity state centrally. Load balancers set it on the channel with UpdateState. Load balancers might as well have a way to read the state back later from the channel when they need it. That avoids having to keep their own copy of the state and worrying about keeping it in sync with the channel.

Since no one else follows this pattern I'll update the load balancers to keep a copy of state. This property will disappear.